doschema/jsvalidation.py

```diff
@@ -0,0 +1,287 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with DoSchema; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Merger module."""
+
+
+import collections
+
+import jsonschema
+
+from doschema.errors import JSONSchemaCompatibilityException, \
+    JSONSchemaDependenciesException, JSONSchemaEnumException, \
+    JSONSchemaRootNotObjectException
+
+
+class JSONSchemaValidation(object):
+    """Class for checking compatibility between schemas."""
+
+    COLLECTION_KEYS = frozenset(['allOf', 'anyOf', 'oneOf'])
+
+    def __init__(self, ignore_index=True, resolver_factory=False):
+        """Constructor.
+
+        :param ignore_index: If set to True, which is default, it will ignore
+                             array indexes.
+        :param resolver_factory: Resolver used to retrieve referenced schemas.
+                                 If not provided it will use default.
+
+        """
+        self.ignore_index = ignore_index
+        self.fields_types_dict = {}
+        self.uri = None
+        self.resolver_factory = resolver_factory or jsonschema.RefResolver
+
+    def start(self, schema, uri):
+        """Start processing a given list of schemas.
+
+        :param schema: Name of schema that is currently processed.
+        :param uri: Name of schema that is currently processed.
+        """
+        self.uri = uri
+        self.resolver = self.resolver_factory(base_uri=uri, referrer=schema)
+
+        self.validate_schema(schema, ())
+
+        self.uri = None
+        self.resolver = None
+
+    def validate_schema(self, curr_schema, curr_field):
+        """Go through the schema and retrieve schema's particular fields.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        self.validate_collection(curr_schema, curr_field)
+
+        while '$ref' in curr_schema:
+            path = curr_schema.get('$ref')
+            curr_schema = self.resolver.resolve(path)[1]
+
+        field = self.fields_types_dict.get(curr_field)
+
+        if "type" in curr_schema:
+            if field and curr_schema['type'] != field.field_type:
+                err_msg = "Type mismatch in schemas {0} and {1}."
+                raise JSONSchemaCompatibilityException(
+                    field.schema_index,
+                    self.uri,
+                    err_msg.format(field.schema_index, self.uri)
+                )
+            self.add_field_to_dict(curr_field, curr_schema['type'])
```

Comment:
- There is no need to add the field type again if it was already registered.
- On the other hand, `field` is used again later but it is not updated after calling `add_field_to_dict`.
- Also `add_field_to_dict ` has only one line and it is not cheap to call a function in python. It might be better to inline this call.
What about this instead?
```diff
-  self.add_field_to_dict(curr_field, curr_schema['type'])
+ if field is not None:
+    field = self.fields_types_dict[curr_field] = FieldToAdd(
+            schema_index=self.uri,
+            field_tuple=curr_field,
+            field_type=curr_schema['type']
+        )
```